practice/038/answer/impl.ml

```diff
@@ -0,0 +1,31 @@
+let rec gcd a b = if b = 0 then a else gcd b (a mod b)
+let coprime a b = gcd a b = 1
+
+let phi n =
+  let rec count_coprime acc d = if d < n then count_coprime (if coprime n d then acc + 1 else acc) (d + 1) else acc in
+  if n = 1 then 1 else count_coprime 0 1
+
+(* Naive power function. *)
+let rec pow n p = if p < 1 then 1 else n * pow n (p - 1)
+
+let factors n =
+  let rec aux d n =
+    if n = 1 then [] else
+      if n mod d = 0 then
+        match aux d (n / d) with
+        | (h, n) :: t when h = d -> (h, n + 1) :: t
+        | l -> (d, 1) :: l
+      else aux (d + 1) n
+  in aux 2 n
+
+let phi_improved n =
+  let rec aux acc = function
+    | [] -> acc
+    | (p, m) :: t -> aux ((p - 1) * pow p (m - 1) * acc) t
+  in aux 1 (factors n)
+
+let timeit f a =
+  let t0 = Unix.gettimeofday () in
+  ignore (f a);
+  let t1 = Unix.gettimeofday () in
+  t1 -. t0
```

Comment:
This should be part of file `ex.ml` not this one, otherwise we ask the system under test to provide its one time measuring procedure, which allows cheating ;-)

Reply:
I understand what you're saying: testing framework should be independent of the system under test. 

But, the [`solution`](https://github.com/ocaml/ocaml.org/blob/main/data/exercises/038_compare_totient.md#solution) is the `timeit` function. 

So should we just put `timeit` in `ex.ml` and then just ask the user to implement `phi` and `phi_improved` again and then just run the tests?

Reply:
These exercises are derived from a Lisp implementation but were initially designed for Prolog. I believe exercise 38 is just not a good idea. It should be part of 37 to check the results are the same as in 34 and the performance is better. Asking the users to write the `timeit` function does not have a great value

Reply:
@cuihtlauac That's true. So should we not do anything for `038`?

Reply:
Yes, 38 should be deleted. I've created this: https://github.com/ocaml/ocaml.org/issues/2278

Reply:
@cuihtlauac Okay, on it 🫡